action/protocol/staking/candidate_center_test.go

```diff
@@ -461,59 +462,106 @@ func TestMultipleNonStakingCandidate(t *testing.T) {
+		dk.Reset()
+		candcenter = candCenterFromNewCandidateStateManager(r, view, dk)
+		r.True(testEqual(candcenter, CandidateList(cands)))
 	}
-	// add candidates
-	r.NoError(candcenter.Upsert(cand))
-	checkCandidates(candcenter, []*Candidate{cand})
-	// self stake candidate
-	candStaked := cand.Clone()
-	candStaked.SelfStakeBucketIdx = 1
-	candStaked.SelfStake = unit.ConvertIotxToRau(1200000)
-	r.NoError(candcenter.Upsert(candStaked))
-	checkCandidates(candcenter, []*Candidate{candStaked})
-	// change self stake bucket index
-	candUpdated := candStaked.Clone()
-	candUpdated.SelfStakeBucketIdx = 2
-	r.NoError(candcenter.Upsert(candUpdated))
-	checkCandidates(candcenter, []*Candidate{candUpdated})
-	r.False(candcenter.ContainsSelfStakingBucket(candStaked.SelfStakeBucketIdx))
+	t.Run("nonstaked candidate not collision on bucket", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		r.NoError(candcenter.Upsert(candNonStaked2))
+		checkCandidates(candcenter, []*Candidate{candNonStaked1, candNonStaked2})
+	})
+	t.Run("staked candidate collision on bucket", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candStaked))
+		r.ErrorIs(candcenter.Upsert(candStakedColBucket), ErrInvalidSelfStkIndex)
+		checkCandidates(candcenter, []*Candidate{candStaked})
+	})
+	t.Run("nonstaked candidate collision on operator", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		r.ErrorIs(candcenter.Upsert(candNonStaked1ColOpt), ErrInvalidOperator)
+		checkCandidates(candcenter, []*Candidate{candNonStaked1})
+	})
+	t.Run("nonstaked candidate collision on name", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		r.ErrorIs(candcenter.Upsert(candNonStaked1ColName), action.ErrInvalidCanName)
+		checkCandidates(candcenter, []*Candidate{candNonStaked1})
+	})
+	t.Run("nonstaked candidate update on owner", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		r.NoError(candcenter.Upsert(candNonStaked1ColOwner))
+		checkCandidates(candcenter, []*Candidate{candNonStaked1ColOwner})
+	})
+	t.Run("change bucket", func(t *testing.T) {
+		candcenter, err := NewCandidateCenter(nil)
+		r.NoError(err)
+		r.NoError(candcenter.Upsert(candNonStaked1))
+		// settle self-stake bucket
+		candStaked := candNonStaked1.Clone()
+		candStaked.SelfStakeBucketIdx = 1
+		candStaked.SelfStake = unit.ConvertIotxToRau(1200000)
+		r.NoError(candcenter.Upsert(candStaked))
+		// change self-stake bucket
+		candUpdated := candStaked.Clone()
+		candUpdated.SelfStakeBucketIdx = 2
+		r.NoError(candcenter.Upsert(candUpdated))
+		checkCandidates(candcenter, []*Candidate{candUpdated})
```

Comment:
`r.False(candcenter.ContainsSelfStakingBucket(candStaked.SelfStakeBucketIdx))`
this is a good test line to keep